### PR TITLE
Week 2 servers

### PR DESCRIPTION
**Pull request by kibazohb**
Pull request for Step 1 & 2 & 3 in week 2 exercise.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -15,6 +15,9 @@
 package com.google.sps.servlets;
 
 import java.io.IOException;
+import java.io.*; 
+import java.util.*;
```
Comment (classm-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Although wildcard imports are terribly convenient, they are prohibited by the style guide: https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

Please explicitly import the symbols that you need

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -26,7 +29,20 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    ArrayList<String> facts = new ArrayList<String>();
+    facts.add("I am from Tanzania");
+    facts.add("I am a huge basketball fan");
+    facts.add("Kevin Durants douple ganger");
+    //String name = "Brian Kibazohi";
```
Comment (classm-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Delete this unused, commented code

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -1,28 +1,29 @@
...
-      ['Hello world!', '¡Hola Mundo!', '你好，世界！', 'Bonjour le monde!'];
+//this is the script js file to get sample data from server
 
-  // Pick a random greeting.
-  const greeting = greetings[Math.floor(Math.random() * greetings.length)];
+function getMyName() {
+  fetch('/data').then(response => response.text()).then((name) => {
+    document.getElementById('name-container').innerText = name;
+  });
+}
```
Comment (classm-sps) [inline portfolio/src/main/webapp/script.js]: Doesn't look like this function is used...remove?

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -1,28 +1,29 @@
...
+
+/** Fetches all facts */
 
-  // Add it to the page.
-  const greetingContainer = document.getElementById('greeting-container');
-  greetingContainer.innerText = greeting;
+function getMyFacts() {
+    fetch("/data").then(response => response.json()).then((facts) => {
+        console.log(facts);
```
Comment (classm-sps) [inline portfolio/src/main/webapp/script.js]: general comment about logging: it's always a good idea to add some context to the log message to help understand what's being logged. Seeing a random log line that says ["blah"] is less helpful than "Facts received from server: ["blah"]".

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -1,28 +1,29 @@
...
+        const factsListElement = document.getElementById('liElement-container');
+        factsListElement.innerHTML = '';
+        for (let i = 0; i < facts.length; i++){
```
Comment (classm-sps) [inline portfolio/src/main/webapp/script.js]: nit: please add a space before the `{`

Comment (classm-sps) [inline portfolio/src/main/webapp/script.js]: given that the index isn't important, consider using a for/of loop: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

**Review by classm-sps — CHANGES_REQUESTED** [review]
Looking great, just a few comments.

File: portfolio/src/main/webapp/script.js (hunk 3 above)
Comment (kibazohb) [reply portfolio/src/main/webapp/script.js]: resolved.

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (kibazohb) [reply portfolio/src/main/webapp/script.js]: resolved.

Comment (kibazohb) [reply portfolio/src/main/webapp/script.js]: I never thought that this would be an issue. Thanks for pointing it out.

Comment (kibazohb) [reply portfolio/src/main/webapp/script.js]: resolved

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -23,10 +32,57 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  ArrayList<String> comments = new ArrayList<String>();
```
Comment (classm-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: something to think about: what are the implications of having `comments` as a member variable vs as a local variable inside `doGet`? Does it matter? Does this work the way you want it to?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -23,10 +32,57 @@
...
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      long id = entity.getKey().getId();
+      String comment = (String) entity.getProperty("Content");
+
+      comments.add(comment);
```
Comment (classm-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:50]: This works, but I'd encourage you to read and understand the example code, where they define a class (`Task` I think) to hold the data coming back from the datastore and pass back to the UI.

**Review by classm-sps — CHANGES_REQUESTED** [review]
You mentioned that datastore wasn't working for you...have you fixed your issues?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 7 above)
Comment (kibazohb) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:50]: I saw how they implemented it in the example, looks like a well constructed solution.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (kibazohb) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I looked into, and understood why I did it. It makes more sense to make it a local variable.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -23,10 +33,60 @@
...
+    String content = getParameter(request, "text-input", "");
+    long timestamp = System.currentTimeMillis();
+    Entity commentEntity = new Entity("Comment");
+    
+    commentEntity.setProperty("Content", content);
+    commentEntity.setProperty("timestamp", timestamp);
+    datastore.put(commentEntity);
+    
+    response.sendRedirect("/blog.html");
+
```
Comment (classm-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:76]: nit: delete this blank line before the closing brace, and put a blank line after the closing brace

**Review by classm-sps — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (kibazohb) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:76]: Sure I will, I appreciate the little things you've noticed.

**Pull request merged**